src/core/common/modulemanager.cpp

```diff
@@ -38,81 +38,121 @@
+
+    auto helper = [](auto& self, std::vector<ModuleContainer>& containers, const std::string& lname,
+                     std::unordered_set<std::string>& visited,
+                     std::unordered_set<std::string>& tmpVisited,
+                     std::vector<std::string>& sorted) -> void {
+        auto it = std::ranges::find(containers, lname, &ModuleContainer::identifier);
+        if (it == containers.end()) {
+            throw Exception(IVW_CONTEXT_CUSTOM("ModuleManager"), "Missing module dependency {}",
+                            lname);
+        }
+
+        if (visited.contains(lname)) return;  // Already visited;
+        if (tmpVisited.contains(lname)) {
+            throw Exception("Dependency graph not a DAG",
+                            IVW_CONTEXT_CUSTOM("TopologicalModuleSort"));
+        }
+
+        tmpVisited.insert(lname);
+        for (const auto& [dependency, version] : it->dependencies()) {
+            self(self, containers, dependency, visited, tmpVisited, sorted);
+        }
+        visited.insert(lname);
+        sorted.push_back(lname);
+    };
+
+    std::unordered_set<std::string> visited;
+    std::unordered_set<std::string> tmpVisited;
+    std::vector<std::string> sorted;
+    for (const auto& item : containers) {
+        helper(helper, containers, item.identifier(), visited, tmpVisited, sorted);
+    }
+    // Sort modules according to dependency graph
+    std::sort(containers.begin(), containers.end(), [&](const auto& a, const auto& b) {
+        return std::ranges::find(sorted, a.identifier()) <
+               std::ranges::find(sorted, b.identifier());
+    });
+}
+
+}  // namespace
+
 ModuleManager::ModuleManager(InviwoApplication* app)
-    : app_{app}
-    , protected_{}
-    , onModulesDidRegister_{}
-    , onModulesWillUnregister_{}
-    , libraryObserver_{app}
-    , sharedLibraries_{}
-    , clearLibs_([&]() {
-        util::reverse_erase_if(sharedLibraries_, [this](const auto& module) {
-            // Figure out module identifier from file name
-            auto moduleName = util::stripModuleFileNameDecoration(module->getFilePath());
-            return !this->isProtected(moduleName);
-        });
-        // Leak the protected dlls here to avoid openglqt crash
-        for (auto& lib : sharedLibraries_) lib->release();
-    })
-    , factoryObjects_{}
-    , modules_{}
-    , clearModules_([&]() {
-        // Need to clear the modules in reverse order since the might depend on each other.
-        // The destruction order of vector is undefined.
-        util::reverse_erase(modules_);
-    }) {}
-
-ModuleManager::~ModuleManager() = default;
+    : app_{app}, onModulesDidRegister_{}, onModulesWillUnregister_{}, inviwoModules_{} {}
 
+ModuleManager::~ModuleManager() {
+    // Need to clear the modules in reverse order since the might depend on each other.
+    // The destruction order of vector is undefined.
+    for (auto& cont : inviwoModules_ | std::views::reverse) {
```

Comment:
```suggestion
ModuleManager::~ModuleManager() {
    // Need to clear the modules in reverse order since they might depend on each other.
    // The destruction order of vector is undefined.
    for (auto& cont : inviwoModules_ | std::views::reverse) {
```